### ISSUE TITLE
System.ArgumentOutOfRangeException: 'end' must not be less than 'start' (Parameter 'end')

### ISSUE THREAD

**Issue by WhitWaldo** [postsharp/Metalama#153]
Here's the stack trace for the exception:

```txt
StreamJsonRpc.RemoteInvocationException: 'end' must not be less than 'start' (Parameter 'end')
   at StreamJsonRpc.JsonRpc.<InvokeCoreAsync>d__151`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Metalama.Framework.DesignTime.VisualStudio.Preview.UserProcessTransformationPreviewService.<PreviewTransformationAsync>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Metalama.Framework.DesignTime.Contracts.Preview.TransformationPreviewServiceExtensions.<PreviewTransformationAsync>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Metalama.Vsx.Preview.PreviewCommand.<ExecuteCoreAsync>d__7.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Metalama.Vsx.Preview.PreviewCommand.<Execute_Source>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Metalama.Vsx.Preview.PreviewCommand.<Execute>d__5.MoveNext()
RPC server exception:
System.ArgumentOutOfRangeException: 'end' must not be less than 'start' (Parameter 'end')
      at Microsoft.CodeAnalysis.Text.TextSpan.FromBounds(Int32 start, Int32 end)
      at Metalama.Framework.Engine.Templating.Mapping.TextMapFile.GetSourceLocation(TextSpan targetSpan)
      at Metalama.Framework.Engine.CompileTime.CompileTimeCompilationBuilder.<>c__DisplayClass24_0.<TryEmit>g__ReportDiagnostics|2(IEnumerable`1 diagnostics)
      at Metalama.Framework.Engine.CompileTime.CompileTimeCompilationBuilder.TryEmit(OutputPaths outputPaths, Compilation compileTimeCompilation, IDiagnosticAdder diagnosticSink, TextMapDirectory textMapDirectory, CancellationToken cancellationToken)
      at Metalama.Framework.Engine.CompileTime.CompileTimeCompilationBuilder.TryGetCompileTimeProjectImpl(ClassifyingCompilationContext compilationContext, ProjectLicenseInfo projectLicenseInfo, IReadOnlyList`1 sourceTreesWithCompileTimeCode, IReadOnlyList`1 referencedProjects, ImmutableArray`1 globalUsings, IDiagnosticAdder diagnosticSink, Boolean cacheOnly, CompileTimeProject& project, CancellationToken cancellationToken)
      at Metalama.Framework.Engine.CompileTime.CompileTimeCompilationBuilder.TryGetCompileTimeProject(ClassifyingCompilationContext compilationContext, ProjectLicenseInfo projectLicenseInfo, IReadOnlyList`1 compileTimeTreesHint, IReadOnlyList`1 referencedProjects, IDiagnosticAdder diagnosticSink, Boolean cacheOnly, CompileTimeProject& project, CancellationToken cancellationToken)
      at Metalama.Framework.Engine.CompileTime.CompileTimeProjectRepository.Builder.TryGetCompileTimeProjectFromCompilation(ClassifyingCompilationContext compilationContext, ProjectLicenseInfo projectLicenseInfo, IReadOnlyList`1 compileTimeTreesHint, IDiagnosticAdder diagnosticSink, Boolean cacheOnly, CancellationToken cancellationToken, CompileTimeProject& compileTimeProject)
      at Metalama.Framework.Engine.CompileTime.CompileTimeProjectRepository.Builder.TryGetCompileTimeProject(MetadataReference reference, IDiagnosticAdder diagnosticSink, Boolean cacheOnly, CancellationToken cancellationToken, CompileTimeProject& referencedProject)
      at Metalama.Framework.Engine.CompileTime.CompileTimeProjectRepository.Builder.TryGetCompileTimeProjectFromCompilation(ClassifyingCompilationContext compilationContext, ProjectLicenseInfo projectLicenseInfo, IReadOnlyList`1 compileTimeTreesHint, IDiagnosticAdder diagnosticSink, Boolean cacheOnly, CancellationToken cancellationToken, CompileTimeProject& compileTimeProject)
      at Metalama.Framework.Engine.CompileTime.CompileTimeProjectRepository.Builder.TryBuild(Compilation compilation, ProjectLicenseInfo projectLicenseInfo, IReadOnlyList`1 compileTimeTreesHint, IDiagnosticAdder diagnosticSink, Boolean cacheOnly, CancellationToken cancellationToken, CompileTimeProjectRepository& loader)
      at Metalama.Framework.Engine.Pipeline.AspectPipeline.TryInitialize(IDiagnosticAdder diagnosticAdder, Compilation compilation, ProjectLicenseInfo projectLicenseInfo, IReadOnlyList`1 compileTimeTreesHint, CancellationToken cancellationToken, AspectPipelineConfiguration& configuration)
      at Metalama.Framework.DesignTime.Pipeline.DesignTimeAspectPipeline.PipelineState.GetConfiguration(PipelineState& state, Compilation compilation, Boolean ignoreStatus, TestableCancellationToken cancellationToken)
      at Metalama.Framework.DesignTime.Pipeline.DesignTimeAspectPipeline.GetConfigurationAsync(PartialCompilation compilation, Boolean ignoreStatus, AsyncExecutionContext executionContext, TestableCancellationToken cancellationToken)
      at Metalama.Framework.DesignTime.Pipeline.DesignTimeAspectPipeline.GetConfigurationAsync(PartialCompilation compilation, Boolean ignoreStatus, AsyncExecutionContext executionContext, TestableCancellationToken cancellationToken)
      at Metalama.Framework.DesignTime.Preview.PreviewPipelineBasedService.PrepareExecutionAsync(ProjectKey projectKey, String syntaxTreeName, TestableCancellationToken cancellationToken)
      at Metalama.Framework.DesignTime.Preview.TransformationPreviewServiceImpl.PreviewTransformationAsync(ProjectKey projectKey, String syntaxTreeName, TestableCancellationToken cancellationToken)
```

Now, unfortunately I have yet to recreate this in a minimally reproduceable project and I cannot yet share the project I do have, so we'll have to make do with describing what I've got going on. 

My class generally looks as follows (anything unmetnioned is commented out at the moment):

```cs
public sealed class MyAttribute : TypeAspect
{
  public override void BuildAspect(IAspectBuilder<INamedType> builder)
  {
    var lockObj = GetLockObject(builder);
    var method = builder.Advice.IntroduceMethod(builder.Target, nameof(DoSomething), IntroductionScope.Default, OverrideStrategy.Override, b => 
      { 
        b.Accessibility = Accessibility.Private; 
      }, 
      args: new 
      { 
        lockObject = lockObj; 
      });
  }

  [Template]
  private void DoSomething([CompileTime] IFieldOrProperty lockObject)
  {
    var lockObj = (object)lockObject.Value;
    lock (lockObj)
    {
    }
  }

  [CompileTime]
  internal static IFieldOrProperty GetLockObject(IAspectBuilder<INamedType> builder)
  {
      var initExpr = ExpressionFactory.Parse("new()");
      return GetObject(builder, typeof(object), "_lockObj", initExpr, Accessibility.PrivateProtected);
  }

  [CompileTime]
  private static IFieldOrProperty GetObject(IAspectBuilder<INamedType> builder, Type expectedType,
      string expectedName, IExpression initializerExpression, Accessibility accessibility = Accessibility.Private)
  {
      var obj = builder.Target.AllFieldsAndProperties.FirstOrDefault(a =>
          a.Type.Is(expectedType) && a.Name == expectedName);

      if (obj != default)
          return obj;

      //Else, introduce and override it
      var targetType = ((INamedType) TypeFactory.GetType(expectedType));
      var field = builder.Advice.IntroduceField(targetType, expectedName, IntroductionScope.Target,
          OverrideStrategy.Override,
          b =>
          {
              b.Accessibility = accessibility;
              b.InitializerExpression = initializerExpression;
          });

      return field.Declaration;
  }
}
```

The type I'm applying the aspect to inherits from an abstract class that defines and initializes `_lockObj` as referred to above:
```cs
protected object _lockObj = new();
```

If there's anything specific I can provide beyond this to help narrow down the issue, please let me know. And if I resolve this, I'll try to illustrate what I changed in hopes that helps as well.

**Comment by PostSharpBot** [comment]
Hello @WhitWaldo, thank you for submitting this issue. We will try to get back to you as soon as possible.
Note to the PostSharp team, this ticket is being tracked in our dashboard under ID TP-33314.

**Comment by WhitWaldo** [comment]
Well, it also may have absolutely nothing to do with the method introduction at all. Now I'm getting that exception no matter what change I make and apply to the aspect. At this point, the entire body of `BuildAspect` is commented out and I still cannot get the error to stop appearing.

Restarting Visual Studio did not help - I still see the error in the bar "Metalama ran into an unexpected problem: 'end' must not be less than 'start' (Parameter 'end')"

**Comment by WhitWaldo** [comment]
Is it safe to delete the entire Metalama folder from %AppData% in hopes I can clear the error? The entire aspect is commented out and I'm still seeing the error when I rebuild the the aspect project.

**Comment by svick** [comment]
It appears that your compile-time code has an error or a warning, and that triggers a Metalama bug that causes the exception.

The code you pasted has this:

```c#
args: new 
{ 
  lockObject = lockObj; 
}
```

Which is a syntax error and is the kind of thing that could confuse Metalama. If that is your actual code, does removing the semicolon fix the exception?

**Comment by WhitWaldo** [comment]
I think the semicolon was a transposition error - I don't see it in the source code.

Unfortunately, it seems I'm entirely unable to do anything in that project at the moment. Even after completely commenting out the entire aspect, any attempt to build the project is throwing that exception now. 

I closed VS and deleted all but the crash reports of /AppData/Local/Temp/Metalama in hopes it'd let me do something in the project again. While I haven't rebooted yet, I did try restarting VS and I see the same error, so that isn't boding well.

**Comment by svick** [comment]
I suspect this is caused by something in the source code, so deleting files or restarting VS probably isn't going to help.

Have you tried commenting out your `[CompileTime]` code?

**Comment by WhitWaldo** [comment]
Working backwards to see what might be causing this, you're right. Commented out a whole file and suddenly it's working again.  

I think it doesn't like me trying to pass the value of a field through the `args` object. If I instead pass in the field as a `[CompileTime]` parameter and then reference it within the introduced method, it works.

**Comment by WhitWaldo** [comment]
Ok, here's the offending code and I'm at a loss to what's wrong with it. In the aspect itself, I do not reference this at all (commented out).

In a separate class then:

```cs
[CompileTime]
internal static class MethodBuilder
{
    internal static IIntroductionAdviceResult<IMethod> Build(IAspectBuilder<INamedType> builder,
        IFieldOrProperty lockObjField)
    {
        var method = builder.Advice.IntroduceMethod(builder.Target, nameof(DoSomething),
            IntroductionScope.Target, OverrideStrategy.Override,
            b =>
            {
                b.Accessibility = Accessibility.Private;
                b.IsStatic = false;
            }, args: new
            {
                lockObjField = lockObjField
            });
        return method;
    }

    [Template]
    private static void DoSomething([CompileTime] IFieldOrProperty lockObjField)
    {
        var lockObj = (object)lockObjField.Value;

        lock (lockObj)
        {
        }
    }
}
```

If I comment this entire file out, it builds without an issue. As soon as I uncomment it and try again there's the same "'end' must not be less than 'start'" exception.

**Comment by svick** [comment]
Thanks, I can reproduce the exception using that code. I will work on a fix.

**Comment by svick** [comment]
Ok, so the problem is having a `static` `[CompileTime]` class. I don't yet understand why that's causing the exception, but removing `static` from the class should work around the issue.

**Comment by WhitWaldo** [comment]
Good to know - I'll file a feature request to have that changed (as again, I don't really want to pollute my aspect class itself with a bunch of templates being referenced by other utilities.

**Comment by WhitWaldo** [comment]
Ok, so I shifted the template and creating method out of a static utility and into the main aspect, but I find that I can reproduce this error again without invoking any static values. 

If the method template has the following signature, it works fine and the diff generates as expected:

```cs
[Template]
private void ProcessDictionaryClearedNotification<[CompileTime] TEntity>(IFieldOrProperty lockField, NotifyDictionaryClearEventArgs<Guid, TEntity> e)
{
}
```

But if I add a parameter that includes a compile-time dictionary that contains all the fields I want to reference, I get the same exception (Metalama ran into an unexpected problem: 'end' must not be less than 'start' (Parameter 'end')":

```cs
[Template]
private void ProcessDictionaryClearedNotification<[CompileTime] TEntity>([CompileTime]Dictionary<string, List<IndexMetadata>> propertyIndexes, IFieldOrProperty lockField, NotifyDictionaryClearEventArgs<Guid, TEntity> e)
{
}
```

Now I filed another ticket about a weird instance of the type being marked as nullable between when I define and initialize it's variable (without a nullable mark) and when I return it from the method (suddenly marked as nullable) in #157, but I don't think that's impacting this here. I see the exception whether I use `var` when calling that helper method (and it gets typed `Dictionary<string, List<IndexMetadata>>?` or explicitly cast it as `Dictionary<string, List<IndexMetadata>>`.

**Comment by svick** [comment]
The exception is masking the actual error, the next build of 2023.1 is going to have a fix for that.

In the meantime, I tried a simple use of your template and it worked for me, so I don't know what the problem is:

```c#
builder.Advice.IntroduceMethod(
    builder.Target, nameof(ProcessDictionaryClearedNotification),
    args: new { TEntity = typeof(object), propertyIndexes = new Dictionary<string, List<IndexMetadata>> { { "foo", new() { new IndexMetadata() } } }, lockField });
```

**Comment by WhitWaldo** [comment]
Well, just tried it again and it once again threw that error, so I guess I'm blocked on this aspect until that patch comes out. Please release it as soon as you're able so I can figure out what the actual underlying problem is here!

**Comment by svick** [comment]
@WhitWaldo The exception is fixed in version 2023.1.8-rc. Can you try again using that? Thanks.

**Comment by WhitWaldo** [comment]
The exception has gone away and been replaced with another error, for which I'll be making an issue shortly.
